text/0000-contingency-plan.md

```diff
@@ -0,0 +1,425 @@
+onerous. We may want to consider a policy for dropping older,
+deprecated type-system rules after some time, as discussed in the
+section on *unresolved questions*.
+
+## Notes on phasing
+
+# Alternatives
+
+**Rather than supporting opt-in changes, one might consider simply
+issuing a new major release for every such change.** Put simply,
+though, issuing a new major release just because we want to have a new
+keyword feels like overkill. This seems like to have two potential
+negative effects. It may simply cause us to not make some of the
+changes we would make otherwise, or work harder to fit them within the
+existing syntactic constraints. It may also serve to dilute the
+meaning of issuing a new major version, since even additive changes
+that do not affect existing code in any meaningful way would result in
+a major release. One would then be tempted to have some *additional*
+numbering scheme, PR blitz, or other means to notify people when a new
+major version is coming that indicates deeper changes.
+
+**Rather than simply fixing soundness bugs, we could use the opt-in
+mechanism to fix them conditionally.** This was initially considered
+as an option, but eventually rejected for the following reasons:
+
+- This would effectively cause a deeper split between minor versions;
+  currently, opt-in is limited to "surface changes" only, but allowing
+  opt-in to affect the type system feels like it would be creating two
+  distinct languages.
+- It seems likely that all users of Rust will want to know that their code
+  is sound and would not want to be working with unsafe constructs or bugs.
+- Users may choose not to opt-in to newer versions because they do not
+  need the new features introduced there or because they wish to
+  preserve compatibility with older compilers. It would be sad for
+  them to lose the benefits of bug fixes as well.
+- We already have several mitigation measures, such as opt-out or
+  temporary deprecation, that can be used to ease the transition
+  around a soundness fix. Moreover, separating out new type rules so
+  that they can be "opted into" can be very difficult and would
+  complicate the compiler internally; it would also make it harder to
+  reason about the type system as a whole.
+
+**Rather than using a version number to opt-in to minor changes, one
+might consider using the existing feature mechanism.** For example,
+one could write `#![feature(foo)]` to opt in to the feature "foo" and
+its associated keywords and type rules, rather than
+`#![rust_version="1.2.3"]`. While using minimum version numbers is
+more opaque than named features, they do offer several advantages:
+
+1. Using a version number alone makes it easy to think about what
+   version of Rust you are using as a conceptual unit, rather than
+   choosing features "a la carte".
+2. Using named features, the list of features that must be attached to
+   Rust code will grow indefinitely, presuming your crate wants to
+   stay up to date.
```

Comment:
@nrc 

>  I hope we don't have so many opt-in features between major versions that this becomes a problem

To expound a bit here, I do think that if we start to accumulate a lot of opt-in changes, it's probably a sign that it's time for a new major version. Using feature names rather than version numbers makes that pressure more obvious, which could be seen as an argument either for or against it.